src/Data/UTxO/Transaction.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+
+module Data.UTxO.Transaction
+    ( UTxOTransaction (..)
+    , N(..)
+    ) where
+
+import Prelude
+
+import Data.ByteString
+    ( ByteString )
+import Data.Word
+    ( Word32 )
+import Numeric.Natural
+    ( Natural )
+
+-- | Construct a UTxO Transaction from inputs and outputs. The interface is
+-- pretty generic and may be a bit _frightening_ but it's in essence very simple
+-- and ensure that the transaction is correct-by-construction.
+--
+-- For instance, cosntructing a transaction with one input and two inputs is as
+-- simple as:
+--
+-- @
+-- let transaction = empty
+--          & addInput   inp0
+--          & addOutput  out0
+--          & addOutput  out1
+--          & lock
+--          & addWitness wit0
+--          & serialize
+-- @
+--
+-- The nice thing about this interface is that it lets client choose underlying
+-- concrete types which remove a lot of the complexity AND, it prevents the
+-- serialization of illed formed transactions with missing inputs.
+class UTxOTransaction dom where
+    type Init    dom :: *
+
+    type Input   dom = inp | inp -> dom
+    type Output  dom = out | out -> dom
+    type SignKey dom = key | key -> dom
+
+    type CoinSel dom = (sel :: N -> N -> *) | sel -> dom
+    type Tx      dom = (tx  :: N -> *)      | tx  -> dom
+
+    -- * Primitive types construction
+    mkInput   :: ByteString -> Word32  -> Maybe (Input dom)
+    mkOutput  :: ByteString -> Natural -> Maybe (Output dom)
+    mkSignKey :: ByteString -> Maybe (SignKey dom)
+
+    -- * Transaction construction
+    empty     :: Init dom   -> CoinSel dom Zero Zero
+    addInput  :: Input dom  -> CoinSel dom n m -> CoinSel dom (Some n) m
+    addOutput :: Output dom -> CoinSel dom n m -> CoinSel dom n (Some m)
+
+    -- * Transaction signature
+    lock       :: CoinSel dom (Some n) (Some m) -> Tx dom (Some n)
```

Comment:
is `lock` really needed? inputs/outputs are going to be very different in terms of payload to witness, right?

Reply:
That `lock` steps prevent from mixing up witnesses and inputs. Let's consider lock was there, then one could write:

```hs
empty & addInput i0 & addWitness w0 & addInput i1 & addWitness w1
```

Which is totally wrong because, in order to compute the witness, we need the transaction id which is a hash of the whole transaction body. Hence, the transaction needs to be complete and have all its inputs and outputs before the first witness can be computed.

Thus yes, the `lock` step is definitely required to prevent the construction of invalid transactions.